daemon,pkg/service: Use hive cell infra for pkg/service

rastislavs opened the pull request:
This PR uses hive cell infra for wiring the `pkg/service` in the Daemon. This allows for easier injection of the `pkg/service` dependency in other modules in the future. 

Also defines the `ServiceManager` interface for the `pkg/service.Service`, to allow for easier mocking of this dependency.

As it was necessary to avoid an import cycle, this also moves `endpointsynchronizer` from `pkg/k8s/watchers` to `pkg/endpointmanager` so that `endpointmanager` does not need to import the top-level `pkg/k8s/watchers`.

rastislavs committed:
daemon,pkg/service: Use hive cell infra for pkg/service.Service

Uses hive cell for wiring the pkg/service in the Daemon.
This allows for easier injection of pkg/service dependency
in other modules in the future.

This also defines the ServiceManager interface for the
pkg/service.Service, to allow for easier mocking of this dependency.

Signed-off-by: Rastislav Szabo <rastislav.szabo@isovalent.com>

joamaki commented on pkg/service/service.go:
```diff
@@ -2104,3 +2100,11 @@ func (s *Service) SyncServicesOnDeviceChange(nodeAddressing types.NodeAddressing
+
+// sendNotification sends an agent notification to the monitor
+func (s *Service) sendNotification(notification monitorAPI.AgentNotifyMessage) error {
+	if option.Config.DryMode {
```
Not sure if need `DryMode` here. The monitor agent cell has an optional dependency to the events map and thus works fine in non-privileged integration tests.

joamaki commented on pkg/service/api/cell.go:
```diff
@@ -0,0 +1,32 @@
+)
+
+// ServiceManagerCell provides access to the datapath Service Manager.
+var ServiceManagerCell = cell.Module(
```
I would just call this `Cell` and use it as `service.Cell` (if you follow my other comment and move it into pkg/service).

rastislavs committed:
endpointsynchronizer: move from pkg/k8s/watchers to pkg/endpointmanager

Moves endpointsynchronizer from pkg/k8s/watchers to pkg/endpointmanager
so that endpointmanager does not need to import top-level pkg/k8s/watchers.
This avoids import loops for endpointmanager dependants.

Signed-off-by: Rastislav Szabo <rastislav.szabo@isovalent.com>
Suggested-by: Jussi Maki <jussi@isovalent.com>

aditighag commented on pkg/service/cell.go:
```diff
@@ -0,0 +1,31 @@
+
+// Cell provides access to the datapath Service Manager.
+var Cell = cell.Module(
+	"datapath-service-manager",
```
nit: Just `service manager` (`load-balancer-service-manager` might've been better, but there is a load-balancer service type in k8s so it may create confusion)

rastislavs replied:
thx, renamed

aditighag commented on pkg/service/types.go:
```diff
@@ -0,0 +1,71 @@
+	"github.com/cilium/cilium/pkg/lock"
+)
+
+// ServiceManager is a datapath service handler. Its main responsibility is to reflect
```
Suggestion: `ServiceManager` provides an interface for service related operations, and it is implemented by ...

rastislavs replied:
thx, reworded

aditighag commented on pkg/service/cell.go:
```diff
@@ -0,0 +1,31 @@
+	"datapath-service-manager",
+	"Datapath Service Manager",
+
+	cell.Provide(newServiceManager),
```
Is there a shutdown handler you can register that's reciprocal to `cell.Provide(newServiceManager)`?

rastislavs replied:
I don't think so - why would we need it?

What is possible though is to implement [lifecycle hooks](https://docs.cilium.io/en/latest/contributing/development/hive/#lifecycle) (`Start` and `Stop`) for the Service if it is needed. In the current code I have not found a need for them, but they can be added at any time.

aditighag reviewed (changes requested):
>daemon,pkg/service: Use hive cell infra for pkg/service.Service

Thanks! Much needed refactoring change that'll simplify `pkg/service` lifecycle management and dependencies. 

(I didn't review the first commit, but I would suggest adding "Suggested-by: Jussi..." tag to the commit description per OSS guidelines.)

aditighag reviewed (approved):
🚢

aditighag reviewed (approved):
Hmm... It looks like GH didn't recognize Jussi and my reviews for the code owners after you moved the PR out of draft. 
I''ll try approving the changes again in case that does the trick.
